common/src/androidTest/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoaderTest.java

```diff
@@ -0,0 +1,178 @@
+
+    @Before
+    public void setUp() throws Exception {
+        // Everything is on clean slate.
+        AuthenticationSettings.INSTANCE.clearSecretKeysForTestCases();
+        AndroidKeyStoreUtil.deleteKey(MOCK_KEY_ALIAS);
+        FileUtil.deleteFile(getKeyFile());
+    }
+
+    private File getKeyFile() {
+        return new File(
+                context.getDir(context.getPackageName(), Context.MODE_PRIVATE),
+                AndroidWrappedKeyLoader.KEY_FILE_PATH);
+    }
+
+    @Test
+    public void testRSAKeyStoreOperations() throws ClientException {
+        testKeyStoreOperation("RSA");
+    }
+
+    private void testKeyStoreOperation(String keyAlgo) throws ClientException {
+        // Write
+        final KeyPair generatedKeyPair = AndroidKeyStoreUtil.generateKeyPair(
+                keyAlgo,
+                getMockKeyPairGeneratorSpec(MOCK_KEY_ALIAS));
+
+        // Read
+        final KeyPair keyPairReadFromKeyStore = AndroidKeyStoreUtil.readKey(MOCK_KEY_ALIAS);
+
+        Assert.assertEquals(generatedKeyPair.getPrivate(), keyPairReadFromKeyStore.getPrivate());
+        Assert.assertEquals(generatedKeyPair.getPublic(), keyPairReadFromKeyStore.getPublic());
+
+        // Delete
+        AndroidKeyStoreUtil.deleteKey(MOCK_KEY_ALIAS);
+
+        // Read again - should be empty.
+        Assert.assertNull(AndroidKeyStoreUtil.readKey(MOCK_KEY_ALIAS));
+    }
+
+    private AlgorithmParameterSpec getMockKeyPairGeneratorSpec(final String alias){
+        final Date startDate = new Date();
+        final Date endDate =  new Date(startDate.getTime() + 1000000);
+
+        return new KeyPairGeneratorSpec.Builder(context)
+                .setAlias(alias)
+                .setSubject(new X500Principal("CN=SOME_CN, OU=SOME_OU"))
+                .setSerialNumber(BigInteger.ONE)
+                .setStartDate(startDate)
+                .setEndDate(endDate)
+                .build();
+    }
+
+    @Test
+    public void testGenerateKey() throws ClientException {
+        final AndroidWrappedKeyLoader keyLoader = new AndroidWrappedKeyLoader(MOCK_KEY_ALIAS, context, null);
+        final SecretKey secretKey = keyLoader.generateRandomKey();
+
+        Assert.assertEquals(AES256KeyLoader.AES_ALGORITHM, secretKey.getAlgorithm());
+    }
+
+    @Test
+    public void testReadKeyDirectly() throws ClientException {
+        final AndroidWrappedKeyLoader keyLoader = initKeyLoaderWithKeyEntry();
+        final SecretKey secretKey = keyLoader.getKey();
+        final SecretKey storedSecretKey = keyLoader.readSecretKeyFromStorage();
+
+        // They're not the same object!
+        Assert.assertNotSame(secretKey, storedSecretKey);
+
+        Assert.assertEquals(AES256KeyLoader.AES_ALGORITHM, secretKey.getAlgorithm());
+        Assert.assertEquals(AES256KeyLoader.AES_ALGORITHM , storedSecretKey.getAlgorithm());
+
+        Assert.assertArrayEquals(secretKey.getEncoded(), storedSecretKey.getEncoded());
```

Comment:
Should these be null?  They can, by contract, return a null (and I hope ours can be made always to do so, eventually).

Reply:
In this case, it should NOT be null.